design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,238 @@
+**Release Target**: TBD | 
+**Status**: _Discussion_ | 
+**Discussion**: TBD
+
+**tl;dr:** _Introduce **Y**et **A**nother **T**esting **T**ool (YATT) that is written and driven 
+primarily by SQL-based syntax to add coverage for test cases that require interleaving statements
+and inserts. Eventually, the goal is to replace the external facing testing tool and maybe QTT tests._
+           
+## Motivation and background
+
+Our existing testing infrastructure is powerful and can handle many of our existing use cases. It,
+however, lacks the flexibility to be useful when testing query upgrades: neither framework supports
+interleaving statements/inserts/asserts. This KLIP proposes a new, imperative, testing tool to
+cover that use case and describes a way to replace the existing tools with this functionality.
+
+The list below enumerates the motivating design principles:
+
+1. the tool should support imperative tests, allowing interleaved statements/inserts/asserts
+1. the tool should allow tests to be written primarily in SQL, avoiding special-syntax directives
+    when possible
+1. the tool should leverage as much of the ksqlDB engine as possible, keeping testing-specific
+    code to a minimum. Whenever we require new functionality for the testing tool, we will first
+    consider supporting it directly in ksqlDB.
+1. the tool should run many tests quickly, with the ability to parallelize test runs
+
+Another minor benefit of a SQL-based testing tool is to allow inline comments describing statements
+and asserts.
+
+### Why Not Reuse?
+
+Why introduce YATT (Yet Another Testing Tool) when we already have two existing options? There
+are a few factors here:
+
+- the primary motivation is practical in nature: We need interleaved testing for the query upgrades 
+    work, and building this into QTT would require essentially an entire rewrite of both the existing
+    tool and the historical test execution while maintaining or migrating all the historical plans. 
+    In order to deliver query upgrades (see KLIP-28) quickly, we will begin work on YATT while 
+    maintaining support for QTT tests.
+- the testing tool and QTT have, over time, diverged significantly from the ksqlDB engine and require
+    lots of custom code (the ksqldb-functional-test module has 11k lines of Java). Specifically, the
+    custom JSON format requires maintaining a parallel, testing-only framework and serde to convert 
+    inputs/outputs to their desired serialization format. Instead, we should leverage the production 
+    SQL expression functionality to produce data.
+- providing a SQL-driven testing tool will mesh better with our product offering, allowing users to
+    write tests without every leaving the "SQL mindset"
+    
+## Scope
+
+This KLIP covers:
+
+- design of the test file format
+- design of the testing-specific directives
+- outline features that need to be implemented in order to replace QTT/testing-tool
+- sketch out an implementation
```

Comment:
I'm not seeing a sketch of how this will be implemented within the KLIP.

Will this be implemented using a fully blown ksqlDB server, running against a (embedded) Kafka, Connect, Schema Registry? Or will it be using the Streams test driver?

If the former, it's likely to be much slower that QTT. Plus, how will the tool support running tests in parallel?  

As discussed offline, ideally, it would support different modes of operation, e.g.
 - test driver vs running against Kafka
 - Using engine directly (like QTT) vs using the rest endpoint (like RQTT)
 - embedded kafka and other services vs  external 

But I'm really not sure this is easily achievable!  So it would be good to know how this initial version will be implemented...

Reply:
**tl;dr** I'm removing this bullet point from the KLIP 😂 

----

Initially I was going to sketch out an implementation, but I think I might leave that for the PR phase because I'd like to do some more experimentation and I don't want to blow up the scope of (both) this KLIP and the deliverable that I need for query upgrade testing.

My intention is actually to see how bad it is with a full embedded kafka deployment and make a decision based on some "benchmarking". This also involves the least amount of test-specific code, and allows us to reuse as much of the production pipeline as possible (both increasing test coverage, and making things like RQTT migration easy).

I'm somewhat skeptical that the kafka portion of the test was what was slowing it down (I can imagine spinning up a new one for each test would be bad, but if we can reuse an existing cluster I can't imagine that kafka would be a true bottleneck given how fast it can process data).

If I'm wrong, and it proves to be a bottleneck, I'll try to implement a chained topology-test-driver implementation.

Reply:
My reason for asking for implementation details in the KLIP is because it will strongly implement what's possible and the possible edge cases and issues.

Kafka is a high-throughput, fault-tolerant system. It's not particularly low-latency.  I'd be very surprised if switching to a real Kafka cluster doesn't increase test run time by an order of magnitude. Though feel free to test it.

QTT already takes a fair while to run. It's also key to the way anyone working on the engine works. For example, if I'm working on engine features I'll generally be debugging a single QTT test repeatedly after each code change, running small sets of files less frequently to find the next issue, and the whole suite occasionally. 

It's not just about how long it takes the full suite to run. Though I still think that will be a lot slower. If switching to Kafka means a single test takes longer to run, due to having to start Kafka, then it will _drastically_ reduce developer productivity.  I believe this was the reason the test driver was written in the first place.

IMHO, I would restrict this KLIP to implementation using the test driver only. This would allow us to replace QTT and the ksqlDB testing tool once YATT is feature complete enough.  BYOK and RQTT can be mentioned as future work only, requiring additional KLIPs.

Reply:
I tested this and you are right :D i'll restrict the implementation to test-driver

Reply:
Updated in the KLIP under `implementation`